designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+
+* `Layout` field (of type string) is used for plugin chain resolution on initialized projects. This will be the default if no plugins are specified for a subcommand.
+* `Plugins` field (of type map[string]interface{}) is used for option plugin configuration that stores configuration information of any plugin.
+
+* So now, where should external plugins be defined in the configuration?
+
+  * I propose that the external plugin should get encoded in the project configuration as a part of the `layout` field.
+    * For e.g., external plugin `helm.sdk.operatorframework.io/v2-alpha` can be specified through the `--plugins` flag for every subcommand and also be defined in the project configuration in the `layout` field for plugin resolution.
+
+Example `PROJECT` file:
+
+```yaml
+version: "3"
+domain: testproject.org
+layout: 
+- go.kubebuilder.io/v3
+- helm.sdk.operatorframework.io/v2-alpha
+plugins:
+  helm.sdk.operatorframework.io/v2-alpha:
+    resources:
+    - domain: testproject.org
+      group: crew
+      kind: Captain
+      version: v2-alpha
+  declarative.go.kubebuilder.io/v1:
+    resources:
+    - domain: testproject.org
+      group: crew
+      kind: FirstMate
+      version: v1
+repo: github.com/test-inc/testproject
+resources:
+- group: crew
+  kind: Captain
+  version: v1
+```
+
+#### What to pass between Kubebuilder and plugins
+
+<!--- TODO(rashmigottipati): This section needs more details on what structure to pass to the plugins. --->
+
+Currently, the `injector` type is defined as below:
+
+```go
+// injector is used to inject certain fields to file templates.
+type injector struct {
+    // config stores the project configuration.
+    config config.Config
+
+    // boilerplate is the copyright comment added at the top of scaffolded files.
+    boilerplate string
+
+    // resource contains the information of the API that is being scaffolded.
+    resource *resource.Resource
+}
+```
+
+One of the plugin hooks allows plugins to set up their own flags. Commong flags (like `--group`, `--version`, `--kind`) are defined by the plugin system itself, and the resource created with those values populated will be passed to the plugins. `create api` and `create webhook` commands would require the GVK being passed as well to the plugins. Currently, `InjectResource` is a required hook for `create api` and `create webhook` commands and the entire resource (which contains the group, version and kind) is being passed to the plugins for these commands.
+
+All the raw flags of the plugins would have to be passed down to the plugins through injector. The plugin specific flags would have to be specified in the project configuration by the user. And the current implmentation is such that, every subcommand can optionally implement the `InjectConfig` hook which passes the project configuration to the plugins and also enable the plugins to modify it if required.
```

Comment:
The plugin specific flags would have to be specified in the project configuration by the user? I don't think we should require this.

Reply:
Agreed, there's no reason to do this. The plugin binary should handle parsing flags internally.